retrieval/discovery/kubernetes/pods.go

```diff
@@ -0,0 +1,221 @@
+)
+
+func (kd *Discovery) podSources(pods map[string]*Pod) []string {
+	var podNames []string
+	for _, pod := range pods {
+		podNames = append(podNames, podSource(pod))
+	}
+	return podNames
+}
+
+func (kd *Discovery) updatePod(pod *Pod, eventType EventType) *config.TargetGroup {
+	kd.podsMu.Lock()
+	defer kd.podsMu.Unlock()
+
+	switch eventType {
+	case deleted:
+		return kd.deletePod(pod)
+	case added, modified:
+		return kd.addPod(pod)
+	}
+	return nil
+}
+
+func (kd *Discovery) getPods() (map[string]*Pod, string, error) {
+	res, err := kd.queryAPIServerPath(podsURL)
+	if err != nil {
+		return nil, "", fmt.Errorf("Unable to list Kubernetes pods: %s", err)
+	}
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		return nil, "", fmt.Errorf("Unable to list Kubernetes pods. Unexpected response: %d %s", res.StatusCode, res.Status)
+	}
+
+	var pods PodList
+	if err := json.NewDecoder(res.Body).Decode(&pods); err != nil {
+		body, _ := ioutil.ReadAll(res.Body)
+		return nil, "", fmt.Errorf("Unable to list Kubernetes pods. Unexpected response body: %s", string(body))
+	}
+
+	podMap := map[string]*Pod{}
+	for idx, pod := range pods.Items {
+	  log.Infof("Got pod %s", pod.ObjectMeta.Name)
+		podMap[pod.ObjectMeta.Name] = &pods.Items[idx]
+	}
+
+	return podMap, pods.ResourceVersion, nil
+}
+
+func (kd *Discovery) watchPods(events chan interface{}, done <-chan struct{}, retryInterval time.Duration) {
+	until(func() {
+		pods, resourceVersion, err := kd.getPods()
+		if err != nil {
+			log.Errorf("Cannot initialize pods collection: %s", err)
+			return
+		}
+
+		kd.pods = map[string]map[string]*Pod{}
+
+		for _, pod := range pods {
+			events <- &podEvent{added, pod}
+		}
+
+		req, err := http.NewRequest("GET", podsURL, nil)
+		if err != nil {
+			log.Errorf("Cannot create pods request: %s", err)
+			return
+		}
+
+		values := req.URL.Query()
+		values.Add("watch", "true")
+		values.Add("resourceVersion", resourceVersion)
+		req.URL.RawQuery = values.Encode()
+		res, err := kd.queryAPIServerReq(req)
+		if err != nil {
+			log.Errorf("Failed to watch pods: %s", err)
+			return
+		}
+		defer res.Body.Close()
+		if res.StatusCode != http.StatusOK {
+			log.Errorf("Failed to watch pods: %d", res.StatusCode)
+			return
+		}
+
+		d := json.NewDecoder(res.Body)
+
+		for {
+			var event podEvent
+			if err := d.Decode(&event); err != nil {
+				log.Errorf("Watch pods unexpectedly closed: %s", err)
+				return
+			}
+
+			select {
+			case events <- &event:
+			case <-done:
+			}
+		}
+	}, retryInterval, done)
+}
+
+func (kd *Discovery) deletePod(pod *Pod) *config.TargetGroup {
+  tg := &config.TargetGroup{Source: podSource(pod)}
+  if _, ok := kd.pods[pod.ObjectMeta.Namespace]; ok {
+    delete(kd.pods[pod.ObjectMeta.Namespace], pod.ObjectMeta.Name)
+    if len(kd.pods[pod.ObjectMeta.Namespace]) == 0 {
+      delete(kd.pods, pod.ObjectMeta.Namespace)
+    }
+  }
+	return tg
+}
+
+func (kd *Discovery) addPod(pod *Pod) *config.TargetGroup {
```

Comment:
Should we filter this to pods that are actually running? Not ready, running. And also check that the pod has an IP address allocated?